velox/benchmarks/basic/LikeFunctionsBenchmark.cpp

```diff
@@ -0,0 +1,202 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+
+#include "vector/fuzzer/VectorFuzzer.h"
+#include "velox/functions/lib/Re2Functions.h"
+#include "velox/functions/lib/benchmarks/FunctionBenchmarkBase.h"
+#include "velox/functions/prestosql/tests/FunctionBaseTest.h"
+#include "velox/tpch/gen/TpchGen.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::tpch;
+using namespace facebook::velox::functions;
+using namespace facebook::velox::functions::test;
+
+DEFINE_int32(vector_size, 10000, "Vector size");
+DEFINE_int32(num_rows, 10000, "Number of rows");
+
+namespace {
+
+class LikeFunctionsBenchmark : public FunctionBaseTest,
+                               public FunctionBenchmarkBase {
+ public:
+  explicit LikeFunctionsBenchmark() {
+    exec::registerStatefulVectorFunction("like", likeSignatures(), makeLike);
+
+    folly::BenchmarkSuspender kSuspender;
+    VectorFuzzer::Options opts;
+    opts.vectorSize = FLAGS_vector_size;
+    VectorFuzzer fuzzer(opts, FunctionBenchmarkBase::pool());
+    inputFuzzer_ = fuzzer.fuzzFlat(VARCHAR());
+    kSuspender.dismiss();
+  }
+
+  // Generate random string using characters from characterSet.
+  std::string generateRandomString(const char* characterSet) {
+    vector_size_t characterSetLength = strlen(characterSet);
+    vector_size_t outputStringLength;
+    vector_size_t minimumLength = 1;
+    vector_size_t maximumLength = 10;
+    outputStringLength = rand() % maximumLength + minimumLength;
+    std::string output;
+
+    for (int i = 0; i < outputStringLength; i++) {
+      output += characterSet[rand() % characterSetLength];
+    }
+    return output;
+  }
+
+  std::string generatePattern(
+      PatternKind patternKind,
+      const std::string& inputString) {
+    switch (patternKind) {
+      case PatternKind::kExactlyN:
+        return std::string(inputString.size(), '_');
+      case PatternKind::kAtLeastN:
+        return generateRandomString(kWildcardCharacterSet);
+      case PatternKind::kPrefix: {
+        auto fixedPatternLength =
+            std::min(vector_size_t(inputString.size()), 10);
+        auto fixedPatternString = inputString.substr(0, fixedPatternLength);
+        return fixedPatternString + generateRandomString(kAnyWildcardCharacter);
+      }
+      case PatternKind::kSuffix: {
+        auto fixedPatternStartIdx =
+            std::max(vector_size_t(inputString.size() - 10), 0);
+        auto fixedPatternString = inputString.substr(fixedPatternStartIdx, 10);
+        return generateRandomString(kAnyWildcardCharacter) + fixedPatternString;
+      }
+      default:
+        return inputString;
+    }
+  }
+
+  size_t run(VectorPtr input, const StringView patternString) {
+    const auto data = makeRowVector({input});
```

Comment:
- pull out data construction and the compilation of the expression out of the measured time into a suspended scope. 
- we found the result to be more accurate and stable (specially oss run) if you run the evaluation in a loop (choose appropriate loop bound based on when you see numbers stabilize) and reuse-vector to avoid measuring allocation time.

ex:
    // DO NOT MEASURE COMPILATION OF QUERY AND INPUT CONSTRUCTION
     folly::BenchmarkSuspender suspender;
    auto scaledData = std::vector<float>();
    scaledData.reserve(kBenchmarkData.size() * scaleFactor_);

    for (int i = 0; i < scaleFactor_; i++) {
      scaledData.insert(
          scaledData.end(), kBenchmarkData.begin(), kBenchmarkData.end());
    }

    auto data =
        vectorMaker_.rowVector({vectorMaker_.flatVector<float>(scaledData)});
    auto exprSet = compile({...});

    SelectivityVector rows(data->size());
    std::vector<VectorPtr> results(exprSet.exprs().size());
    exec::EvalCtx evalCtx(&execCtx_, &exprSet, data.get());

    suspender.dismiss();
    // RUN 10K iterations, for your use case since its string operation probably less is suffieicent.

    size_t cnt = 0;
    for (auto i = 0; i < times * 1'0000; i++) {
      exprSet.eval(rows, evalCtx, results);
      cnt += results[0]->size();
    }
    return cnt;
    ```

Reply:
Thanks for the suggestion, modified accordingly